src/core/value.rs

```diff
@@ -127,6 +131,36 @@ impl Value {
         )
     }
 
+    /// The notation X^<2 is notation for the type (S X)
+    /// The notation X^<(2*n) is notation for the type S (X^n) * X^<n
+    /// Cannot represent >= 2**16 bytes 0 <= n < 16.
+    ///
+    /// # Panics:
+    ///
+    /// Panics if the length of the slice is >= 2^(n + 1) bytes
+    pub fn var_len_buf_from_slice(v: &[u8], mut n: usize) -> Result<Value, Error> {
+        assert!(n < 16);
+        assert!(v.len() < (1 << (n + 1)));
+        let mut iter = BitIter::new(v.iter().copied());
+        let types = Type::powers_of_two_vec(n); // size n + 1
+        let mut res = None;
+        while n > 0 {
+            let v = if v.len() >= (1 << (n + 1)) {
+                let ty = &types[n];
+                let val = decode::decode_value(ty, &mut iter)?;
+                Value::SumR(Box::new(val))
+            } else {
+                Value::SumL(Box::new(Value::Unit))
+            };
+            match res {
```

Comment:
In 336740d2067f77f868d02823e799a448f811c139:

Should pull out the `res =`, as
```
res = match {
    Some(prod) => Some(Value::Prod(Box::new(prod), Box::new(v)),
    None => Some(v)
};
```
or even
```
res = Some(res.map_or_else(
    || Some(v),
    |prod| Value::Prod(Box::new(prod), Box::new(v))),
));
```

Reply:
Actually, on further reading, I think it would be easier to just early-return when `n == 0`, then `res` doesn't need to be an option anymore and the final `unwrap_or` can be dropped.

Reply:
Initializing `res` would essentially be unrolling the first iteration of the loop. Not really sure if that would be better.

Reply:
Ah, yeah, I see. Agreed, probably not better to unroll it.